source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,407 @@
+        const ref Enrollment enroll) dg) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set");
+
+        foreach (row; results)
+        {
+            auto hash = *cast(Hash*)row.peek!(ubyte[])(0).ptr;
+            auto enroll = deserializeFull!Enrollment(row.peek!(ubyte[])(1));
+
+            if (auto ret = dg(hash, enroll))
+                return ret;
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Remove the enrollment data with the given key from the validator set
+
+        Params:
+            enroll_hash = key for an enrollment data to remove
+
+    ***************************************************************************/
+
+    public void removeEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        this.db.execute("DELETE FROM validator_set WHERE key = ?", enroll_hash[]);
+    }
+
+    /***************************************************************************
+
+        Make an enrollment data for enrollment process
+
+        Params:
+            frozen_utxo_hash = the hash of a frozen UTXO used to identify 
+                                a validator and to generate a siging key
+
+        Returns:
+            an enrollment data
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        this.data = new Enrollment();
+
+        // K, frozen UTXO hash
+        this.data.utxo_key = frozen_utxo_hash;
+
+        // N, cycle length
+        this.data.cycle_length = 1008; // freezing period / 2
+
+        // generate random seed value
+        this.random_seed_src = Scalar.random();
+
+        // X, nth image of random seed
+        this.data.random_seed = hashFull(this.random_seed_src);
+        foreach (i; 0 .. this.data.cycle_length-1)
+            this.data.random_seed = hashFull(this.data.random_seed);
+
+        // R, signature noise
+        this.signature_noise = Pair.random();
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        data.enroll_sig = signHash(this.node_key_pair.secret, signature_noise, hash);
+
+        // save information for enrollment data
+        saveNodeEnrollmentData();
+
+        return *this.data;
+    }
+
+    /***************************************************************************
+
+        Check if the enrollment data exists in the validator set.
+
+        Params:
+            enroll_hash = key for an enrollment data which is hash of frozen UTXO 
+
+        Returns:
+            true if the enrollment manager has the enrollment data
+
+    ***************************************************************************/
+
+    public bool hasEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set " ~
```

Comment:
There is an `EXISTS` statement in SQL to allow for short-circuiting, otherwise the query planner execute over the whole table even if it can find a record:
```sql
SELECT EXISTS(SELECT 1 FROM validator_set WHERE key = ?)
```